resource-providers/src/main/java/io/opentelemetry/resource/spring/SpringBootServiceNameGuesser.java

```diff
@@ -0,0 +1,255 @@
+ */
+
+package io.opentelemetry.resource.spring;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+
+/**
+ * A ResourceProvider that will attempt to guess the application name for a Spring Boot service.
+ * When successful, it will return a Resource that has the service name attribute populated with the
+ * name of the Spring Boot application. It uses the following strategies, and the first successful
+ * strategy wins:
+ *
+ * <ul>
+ *   <li>Check for the SPRING_APPLICATION_NAME environment variable
+ *   <li>Check for spring.application.name system property
+ *   <li>Check for application.properties file on the classpath
+ *   <li>Check for application.properties in the current working dir
+ *   <li>Check for application.yml on the classpath
+ *   <li>Check for application.yml in the current working dir
+ *   <li>Check for --spring.application.name program argument (not jvm arg) via ProcessHandle
+ *   <li>Check for --spring.application.name program argument via sun.java.command system property
+ * </ul>
+ */
+@AutoService(ResourceProvider.class)
+public class SpringBootServiceNameGuesser implements ResourceProvider {
+
+  private static final Pattern COMMANDLINE_PATTERN =
+      Pattern.compile(".*--spring\\.application\\.name=([a-zA-Z.\\-_]+).*");
+  private static final Logger logger = LoggerFactory.getLogger(SpringBootServiceNameGuesser.class);
+  private final SystemHelper system;
+
+  public SpringBootServiceNameGuesser() {
+    this(new SystemHelper());
+  }
+
+  // Exists for testing
+  SpringBootServiceNameGuesser(SystemHelper system) {
+    this.system = system;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+
+    logger.debug("Performing Spring Boot service name auto-detection...");
+    Stream<Supplier<String>> finders =
+        Stream.of(
+            this::findByEnvironmentVariable,
+            this::findBySystemProperties,
+            this::findByClasspathApplicationProperties,
+            this::findByCurrentDirectoryApplicationProperties,
+            this::findByClasspathApplicationYaml,
+            this::findByCurrentDirectoryApplicationYaml,
+            this::findByCommandlineArgument);
```

Comment:
I believe you need to adjust the order of these. For example command line arguments should have higher priority than properties file.

Reply:
This is a good point -- I hadn't considered the user/ergonomics of this ordering decision. I had attempted to guess at the most common/frequent things first...but you're totally right, a user that bothers to put stuff on the commandline or environment definitely wants to override what might be on the classpath or filesystem. Will reorder.